docs/book/src/tasks/experimental-features/runtime-sdk/implement-topology-mutation-hook.md

```diff
@@ -35,11 +38,103 @@ External patches have the following advantages:
+    # if metadata.generation isn't the same as observedGeneration Cluster using the ClusterClass should not reconcile.
+    observedGeneration: xx
+    # variables contains a list of all variable definitions, both inline and external, that belong to the ClusterClass.
+    variables: 
+        - name: http-proxy
+          schema:
+            openAPIV3Schema:
+             type: string
+             # default is different so this variable does not match the inline definition.
+             default: "proxy.example3.com"
+             example: "proxy.example.com"
+             description: "proxy for http calls."
+          # namespace is added as this variable's schema isn't equivalent to the inline definition.
+          namespace: lbImageRepository
+          # A variable can have multiple origins (altenate name /sources/??)
+          # TODO: What purpose does this serve?
+         # origins: 
+         #  VariableDiscovery:
+         #  - patch: lbImageRepository # not sure if to use this or variableDiscovery << to list
+        - name: http-proxy
+          schema:
+          openAPIV3Schema:
+            type: string
+            default: "proxy.example.com"
+            example: "proxy.example.com"
+            description: "proxy for http calls."
+          namespace: inline
+```
+
+### Variable namespacing
+Variable definitions can be inline in the ClusterClass or from any number of external DiscoverVariables hooks.
+If all variables that share a name have equivalent schemas the variables are considered to be in the global namespace.
+Note: This is always the case if there is only one variable with a particular name.
+The CAPI components will consider variable definitions to be equivalent when they share a name and `openAPIV3Schema` fields
+excluding `description` and `example` are exactly equal. (TODO: Clarify this.)
+
+If two or more variables which share a name do not have equivalent schemas, the variables will be namespaced. This means they will
+appear namespaced in the ClusterClass `.status.variables` and setting values in the Cluster `.topology.spec.variables` will also require setting a namespace.
+The namespace on a variable will depend on its source (TODO: RUNTIME_EXTENSION name or patch name?)
+Note: When namespaced inline variables can be set without a namespace as they are always in the global namespace.
+
+### Setting values for variables in the Cluster
+Setting variables using external variables requires attention to be paid to variable namespacing, as exposed in the ClusterClass status. 
+Variable values are set in Cluster `.spec.topology.variables`.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: Cluster
+#metadata 
+spec:
+    topology:
+        variables:
+        - name: http_proxy
+          # If not defined or empty this variable is in the global namepsace.
+          # is a list 
+          namespaces: 
+            - lbImageRepository
+            - inline
```

Comment:
kind of thinking if it is better to have a list or a single value
also, wondering if/how we can make this list SSA-friendly (what is the listMapKey?)

Reply:
Do you mean to make the the variables or the namespaces list SSA-friendly?

> Only applicable when +listType=map. A list of field names whose values uniquely identify entries in the list. While there can be multiple keys, listMapKey is singular because keys need to be specified individually in the Go type. The key fields must be scalars.

If we use "namespace" instead of "namespaces" it would be a scalar. But I'm not sure if the list map keys can use a field that is optional (I assume namespace or namespaces is optional).